Unnecessary fetch health permissions

klnfreedom opened the pull request:
This pull request introduces improved permission handling for health data retrieval on Android and adds support for tracking the raw workout activity type in the Dart data model. The changes focus on ensuring that sensitive health data is only accessed when the appropriate permissions are granted, with clear logging and error handling, and on making the workout data model more robust and informative.

**Android: Enhanced permission checks and error handling**
- Added a new `HealthPermissionChecker` class to centralize and simplify permission checks for location, distance, calories burned, and steps data. This class ensures that data is only accessed if the corresponding permissions are granted, improving security and user privacy. (`android/src/main/kotlin/cachet/plugins/health/HealthPermissionChecker.kt`)
- Updated `HealthDataReader` to use `HealthPermissionChecker` for checking permissions before retrieving distance, calories burned, and steps data. If permissions are missing, the code now logs a message and skips data retrieval for that type, instead of attempting access. (`android/src/main/kotlin/cachet/plugins/health/HealthDataReader.kt`) [[1]](diffhunk://#diff-e19d366033a576bea836f4f48d1ea65afa799bb88103ba1d16fdbbbe07373e02R33) [[2]](diffhunk://#diff-e19d366033a576bea836f4f48d1ea65afa799bb88103ba1d16fdbbbe07373e02R404-R426) [[3]](diffhunk://#diff-e19d366033a576bea836f4f48d1ea65afa799bb88103ba1d16fdbbbe07373e02L416-R448) [[4]](diffhunk://#diff-e19d366033a576bea836f4f48d1ea65afa799bb88103ba1d16fdbbbe07373e02L431-R469)
- Improved error handling in `HealthDataReader` by returning a more descriptive error with an empty list when data is unavailable, instead of just an empty list. (`android/src/main/kotlin/cachet/plugins/health/HealthDataReader.kt`)

SamadiPour commented on android/src/main/kotlin/cachet/plugins/health/HealthDataReader.kt:
```diff
@@ -387,7 +389,7 @@ class HealthDataReader(
-            
+
             // Get distance data
             val distanceRequest = healthConnectClient.readRecords(
```
I think you forgot to remove this. You are requesting distance twice.

SamadiPour reviewed (commented):
Thanks for this PR. I also have this issue because it doesn't work without requesting these permissions. You also can't submit to Google Play if you request extra permissions that you don't actually need.
I added a few small notes that can improve it.

theolm commented on android/src/main/kotlin/cachet/plugins/health/HealthDataReader.kt:
```diff
@@ -399,38 +401,70 @@ class HealthDataReader(
-                    ),
-                ),
-            )
             var totalEnergyBurned = 0.0
```
This is set to null on line 481.

klnfreedom commented:
Thanks for the review, @SamadiPour @theolm. I’ve pushed updates resolving the comments:
- Removed the duplicate distance request
- Made distance permission check only require READ_DISTANCE
- Adjusted null handling for totals

klnfreedom commented:
Hi @iarata,

I've created this PR to fix an issue where the app crashes on Android if READ_DISTANCE permissions aren't granted.

Currently, the package assumes these permissions are always present, which causes a crash even if the app doesn't actually need to track distance. This also makes it difficult to pass Play Store reviews due to unnecessary permission requests.

This change allows developers to use the package without being forced to request sensitive permissions they don't use. Could you please review this when you have a chance?

klnfreedom commented:
@iarata Any updates for my PR?
